src/mds/MDBalancer.cc

```diff
@@ -83,14 +83,99 @@ MDBalancer::MDBalancer(MDSRank *m, Messenger *msgr, MonClient *monc) :
 {
   bal_fragment_dirs = g_conf().get_val<bool>("mds_bal_fragment_dirs");
   bal_fragment_interval = g_conf().get_val<int64_t>("mds_bal_fragment_interval");
+  last_bal_rank_mask = "0x0";
+  last_num_mdss = 0;
 }
 
 void MDBalancer::handle_conf_change(const std::set<std::string>& changed, const MDSMap& mds_map)
 {
-  if (changed.count("mds_bal_fragment_dirs"))
+  if (changed.count("mds_bal_fragment_dirs")) {
     bal_fragment_dirs = g_conf().get_val<bool>("mds_bal_fragment_dirs");
-  if (changed.count("mds_bal_fragment_interval"))
+  }
+  if (changed.count("mds_bal_fragment_interval")) {
     bal_fragment_interval = g_conf().get_val<int64_t>("mds_bal_fragment_interval");
+  }
+}
+
+void MDBalancer::handle_rank_mask_bits()
+{
+  if (mds->mdsmap->get_bal_rank_mask() == last_bal_rank_mask &&
+      mds->get_mds_map()->get_num_in_mds() == last_num_mdss) {
+    return;
+  }
+  last_bal_rank_mask = mds->mdsmap->get_bal_rank_mask();
+  last_num_mdss = mds->get_mds_map()->get_num_in_mds();
+
+  std::string bal_hex_str;
+  bal_hex_str.resize(last_bal_rank_mask.size());
+  std::transform(last_bal_rank_mask.begin(), last_bal_rank_mask.end(), bal_hex_str.begin(), ::tolower);
+
+  bool valid_hex = false;
+  num_mdss_in_rank_mask = 0;
+
+  if (bal_hex_str.substr(0, 2) == "0x") {
+    uint32_t quatet_sum = 0;
+
+    reverse(bal_hex_str.begin(), bal_hex_str.end());
+    bal_hex_str.resize(bal_hex_str.size()-2);
+
+    for (uint32_t qpos = 0; qpos < bal_hex_str.size(); qpos++) {
+      if (!isxdigit(bal_hex_str[qpos])) {
+        quatet_sum = 0;
+        break;
+      }
+
+      uint32_t quatet_value;
+      uint32_t offset = 0;
+
+      quatet_value = stoul(bal_hex_str.substr(qpos, 1), nullptr, 16);
+      quatet_sum += quatet_value;
+
+      while (offset < 4) {
+        if (quatet_value & (1 << offset)) {
+          mds_rank_t masked_rank = qpos * 4 + offset;
+          if (masked_rank < (mds_rank_t)last_num_mdss) {
+            num_mdss_in_rank_mask++;
+          }
+        }
+        offset++;
+      }
+    }
+
+    if (quatet_sum && num_mdss_in_rank_mask) {
+      valid_hex = true;
+    }
+  }
+
+  dout(2) << "mds_bal_rank_mask: " << bal_rank_mask << " (validity: " << valid_hex << ")" << dendl;
+
+  if (valid_hex == false) {
```

Comment:
Why not require valid and well-formed hex when changing the config in MDSMap?

Reply:
If the user inputs the hexadecimal number correctly, the hexadecimal string validation logic is not necessary. However, if you enter the wrong hexadecimal number, the balancer may work incorrectly. For example, users give non-hexadecimal characters, or users input values outside the MAX_MDS. Do you think we don't need the hex validation logic?

Reply:
I think it can be a mix of both. Probably MDSMonitor and MDBalancer can use the same validation method in MDSMap.

Reply:
as suggested in https://github.com/ceph/ceph/pull/43284#discussion_r973207144

Reply:
Okay. I will try to simpify them.